src/pylint_datetime/__init__.py

```diff
@@ -1,4 +1,185 @@
-# License: MIT
-# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+"""Pylint checker ensure correct usage of datetime module, especially around aware/naive objects."""
 
-"""checker to ensure correct datetime usage with aware objects in datetime module."""
+from typing import Any
+
+import astroid  # type: ignore
+import pylint  # type: ignore
+from pylint.checkers import BaseChecker  # type: ignore
+
+# function calls to fromisoformat, strptime and fromisoformat can't be checked - they parse strings
+# if the specified timezone is None, still allowed although they produce naive objects
+
+
+class DatetimeChecker(BaseChecker):  # type: ignore
+    """class for custom pylint checker."""
+
+    name = "pylint-datetime"
+    priority = -1
+    msgs = {
+        "W9999": (
+            "timedelta() called without keyword arguments",
+            "timedelta-no-keyword-args",
+            "timedelta() called without keyword arguments.",
+        ),
+        "W9998": (
+            'Function call to "%s"/"replace" should be called with a timezone argument',
+            "timezone-no-argument",
+            'Function call to "%s"/"replace" should be called with a timezone argument',
+        ),
+        "W9997": (
+            'Function call to "%s" can only produce naive objects',
+            "naive-datetime-call",
+            'Function call to "%s" can only produce naive objects, aware objects necessary',
+        ),
+        "W9996": (
+            'Attribute access "%s" should be followed by a call to replace with timezone argument',
+            "naive-datetime-no-timezone",
+            'Attribute access "%s" should be followed by a call to replace with timezone argument',
+        ),
```

Comment:
We should probably use something else as codes, it is not trivial to know what to use as there seem to be no "central authority" to assign codes. The [documentation](https://pylint.pycqa.org/en/latest/development_guide/how_tos/custom_checkers.html#defining-a-message) is also pretty vague. Probably the most relevant part is:

> The `message-id` should be a 4-digit number, prefixed with a message category. There are multiple message categories, these being `C`, `W`, `E`, `F`, `R`, standing for `Convention`, `Warning`, `Error`, `Fatal` and `Refactoring`. The 4 digits should not conflict with existing checkers and the first 2 digits should consistent across the checker (except shared messages).

Using this command we can get a list of all currently available warnings:

```sh
pylint --list-msgs | grep '(W' | sed 's/^:.* (\(W.*\)):.*$/\1/' | sort
```

The current maximum is W4905. Maybe is not a bad idea to put it in the `W9XXX` namespace to make sure it is unlikely that a built-in check matches, but `W99XX` seems like it might be used by others if it was the default suggested by ChatGPT :laughing: 

Looking into some other random `pylint` extensions:

* [`pylint-print`](https://gitlab.com/nwmitchell/pylint-print/-/blob/main/pylint_print/pylint_print.py?ref_type=heads#L27): Uses `W82XX`.
* [`pylint-protobuf](https://github.com/nelfin/pylint-protobuf/blob/dbdfaecfbf248ccc75f6cdef4b29e4fb6a21e70e/pylint_protobuf/__init__.py#L19): Uses `E59XX`, but they parametrize the first 2 digits so it is easy to update if there is a new overlap, which looks like a good idea.
* [`pylint-beam`](https://github.com/kvudata/pylint-beam/blob/main/pylint_beam/checkers.py#L16): Uses `W50XX`.
* [`pylint-google-style-guide-imports-enforcing`](https://github.com/Enforcer/pylint_google_style_guide_imports_enforcing/blob/ff47ef50b26a6d4d7d027997b2e128bd21adfbbb/pylint_google_style_guide_imports_enforcing/__init__.py#L13): Uses `C51XX`.

Seems pretty random. For the lack of a better idea I suggest using:

```py
>>> c = 0
>>> for i in "Frequenz":
...     c += ord(i)
... 
>>> print(c % 10 + 90)
98
```

So `W98XX` (starting at `W9800`)  :grimacing: